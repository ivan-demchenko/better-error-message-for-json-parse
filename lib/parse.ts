import {
  showFancySyntaxException
} from './helper';

export const safeJsonParse = <A>(raw: string, reviver?: (key: any, value: any) => any): A => {
  try {
    return JSON.parse(raw, reviver);
  } catch (e) {
    throw new SyntaxError(showFancySyntaxException(raw, e));
  }
}
