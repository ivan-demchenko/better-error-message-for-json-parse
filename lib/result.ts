export type Result<F, S> = Failure<F> | Success<S>;

export class Failure<F> {
  constructor(public value: F) {}
}

export class Success<S> {
  constructor(public value: S) {}
  map<T>(fn: (val: S) => T): Success<T> {
    return new Success(fn(this.value));
  }
}

export const failure = <F>(reason: F) => new Failure<F>(reason);

export const success = <S>(value: S) => new Success<S>(value);

export const map = <T, P>(fn: (x: T) => P, r: Result<any, T>): Result<any, P> => {
  if (r instanceof Failure) {
    return r;
  }
  return r.map(fn);
}

export const flatMap = <F, T, P>(fn: (x:T) => Result<F, P>, r: Result<F, T>): Result<F, P> => {
  if (r instanceof Failure) {
    return r;
  }
  return fn(r.value);
}

export const alt = <F1, S1, F2, S2>(r1: Result<F1, S1>, r2: Result<F2, S2>): Result<F1, S1> | Result<F2, S2> => {
  if (r1 instanceof Failure) {
    return r2;
  }
  return r1;
}

export default {
  failure,
  success,
  map,
  flatMap,
  alt
}
