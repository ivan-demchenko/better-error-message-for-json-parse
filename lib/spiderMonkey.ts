import {
  OptErrorPlace,
  buildErrorFromJsonAndPosition
} from './helper';

export const extractErrorPositionFromErrorMsg = (msg: string): OptErrorPlace => {
  const res = /at line (\d+) column (\d+)/.exec(msg);
  return res
    ? {
        l: parseInt(res[1], 10) - 1,
        c: parseInt(res[2], 10) - 1
      }
    : null;
}

export const showFancySyntaxException = (rawJson: string, e: SyntaxError): string => {
  const context: Array<string> = [ e.message ];
  const lines = rawJson.split('\n');
  const position = extractErrorPositionFromErrorMsg(e.message);
  return position
    ? buildErrorFromJsonAndPosition(lines, context, position)
    : e.message;
}

export const safeJsonParse = <A>(raw: string, reviver?: (key: any, value: any) => any): A => {
  try {
    return JSON.parse(raw, reviver);
  } catch (e) {
    throw new SyntaxError(showFancySyntaxException(raw, e));
  }
}
