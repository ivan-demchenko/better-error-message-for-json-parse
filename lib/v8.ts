import {
  ErrorPlace,
  buildErrorFromJsonAndPosition
} from './helper';

export const extractErrorPositionFromErrorMsg = (msg: string): number | null => {
  const res = /Unexpected string in JSON at position (\d+)/.exec(msg);
  return res ? parseInt(res[1], 10) : null;
}

export const findErrCoords = (ls: string[], errorProximity: number, lineIdx = 0): ErrorPlace => {
  return errorProximity > ls[lineIdx].length
    ? findErrCoords(ls, errorProximity - ls[lineIdx].length - 1, ++lineIdx)
    : {l: lineIdx, c: errorProximity};
}

export const showFancySyntaxException = (rawJson: string, e: SyntaxError): string => {
  const context: Array<string> = [ e.message ];
  const lines = rawJson.split('\n');
  const absPos = extractErrorPositionFromErrorMsg(e.message);
  if (absPos) {
    const errCoords = findErrCoords(lines, absPos);
    return buildErrorFromJsonAndPosition(lines, context, errCoords);
  }
  return e.message;
}

export const safeJsonParse = <A>(raw: string, reviver?: (key: any, value: any) => any): A => {
  try {
    return JSON.parse(raw, reviver);
  } catch (e) {
    throw new SyntaxError(showFancySyntaxException(raw, e));
  }
}
