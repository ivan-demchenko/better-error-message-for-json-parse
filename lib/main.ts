import { safeJsonParse } from './parse';

export default safeJsonParse;
