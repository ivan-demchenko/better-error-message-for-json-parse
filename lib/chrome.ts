import { mkArray, buildErrorPointer } from './helper';

export const extractErrorPositionFromErrorMsg = (msg: string): number | null => {
  const res = /Unexpected string in JSON at position (\d+)/.exec(msg);
  return res ? parseInt(res[1], 10) : null;
}

export const getErrorPlaceContext = (rawJson: string, position: number): string =>
  rawJson.substring(position - 20, position + 20);

export const buildMessageForSyntaxException = (rawJson: string, e: SyntaxError): string => {
  const context: Array<string> = [];

  const pos = extractErrorPositionFromErrorMsg(e.message);
  if (pos) {
    context.push(getErrorPlaceContext(rawJson, pos));
    context.push(buildErrorPointer(pos));
  }

  return [e.message, context.join('\n')].join('\n');
}

export const safeJsonParse = <A>(raw: string): A => {
  try {
    return JSON.parse(raw);
  } catch (e) {
    throw new SyntaxError(buildMessageForSyntaxException(raw, e));
  }
}
