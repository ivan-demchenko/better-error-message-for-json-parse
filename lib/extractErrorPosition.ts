import { ErrorPlace } from './helper';
import r, {Result} from './result';

const parseLineColumnMsg = (msg: string): Result<string, string[]> => {
  const result = /at line (\d+) column (\d+)/.exec(msg);
  return (result ? r.success([result[1], result[2]]) : r.failure('Parsing of error message failed'))
}

const extractFromPositionMsg = (msg: string, ls: string[]): Result<string, number> => {
  const result = /Unexpected string in JSON at position (\d+)/.exec(msg);
  return result ? r.success(parseInt(result[1], 10)) : r.failure('Parsing of error message failed');
}

const mapCoordsToErrorPlace = (coords: string[]): Result<string, ErrorPlace> => {
  if (coords.length !== 2) {
    return r.failure('Wrong coordinates');
  }
  return r.success({
    l: parseInt(coords[0], 10) - 1,
    c: parseInt(coords[1], 10) - 1
  });
}

export const findErrCoords = (
  lines: string[],
  errorProximity: number,
  lineIdx = 0
): ErrorPlace => {
  return errorProximity > lines[lineIdx].length
    ? findErrCoords(lines, errorProximity - lines[lineIdx].length - 1, ++lineIdx)
    : {l: lineIdx, c: errorProximity};
}

export const errMsgRegExps = (msg: string, ls: string[]): Result<string, ErrorPlace> =>
  r.alt(
    r.flatMap(mapCoordsToErrorPlace, parseLineColumnMsg(msg)),
    r.map((pos: number) => findErrCoords(ls, pos), extractFromPositionMsg(msg, ls))
  )
