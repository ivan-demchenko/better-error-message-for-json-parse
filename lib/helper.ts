declare global {
  // Safari
  interface SyntaxError {
    line: number,
    column: number
  }
}

import { errMsgRegExps } from './extractErrorPosition';
import { Success } from './result';

export type ErrorPlace = {l: number, c: number};

export const mkArray = (lim: number): Array<number> =>
  Array.from(Array(lim), (_, i) => i);

export const buildErrorPointer = (pos: number): string =>
  mkArray(pos - 1).map(() => '-').join('') + '^';

export const showFancySyntaxException = (rawJson: string, e: SyntaxError): string => {
  const context: Array<string> = [ e.message ];
  const lines = rawJson.split('\n');
  const position = errMsgRegExps(e.message, lines);
  return position instanceof Success
    ? buildErrorFromJsonAndPosition(lines, context, position.value)
    : e.message;
}

export const buildErrorFromJsonAndPosition = (
  lines: string[],
  context: string[],
  position: ErrorPlace
): string => {
  if (lines[position.l - 1]) {
    context.push(lines[position.l - 1]);
  }
  context.push(lines[position.l]);
  context.push(buildErrorPointer(position.c));
  if (lines[position.l + 1]) {
    context.push(lines[position.l + 1]);
  }
  return context.join('\n');
}
