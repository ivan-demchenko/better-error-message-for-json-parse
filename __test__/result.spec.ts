import { join } from 'path';
import { readFileSync } from 'fs';
import { showFancySyntaxException } from '../lib/helper';

const readFixture = (fixtureName: string): string =>
  readFileSync(join(__dirname, 'fixtures', fixtureName)).toString();

describe('building message', () => {
  it('case 1', () => {
    const json = readFixture('test1.json');
    const err = new SyntaxError('Unexpected string in JSON at position 118');

    const expected = [
      'Unexpected string in JSON at position 118',
      '  "surname": "Rowland"',
      '  "fullname": "Denise Pickett",',
      '-^',
      '  "email": "jack@garrison.vc",'
      ].join('\n');

    // then
    const result = showFancySyntaxException(json, err)
    expect(result).toEqual(expected);
    expect(result).toMatchSnapshot();
  });

  it('case 2', () => {
    const json = readFixture('test2.json');
    const err = new SyntaxError('Unexpected string in JSON at position 105');
    
    const expected = [
      'Unexpected string in JSON at position 105',
      '  "name": "Eva",',
      '  "surname" "Rowland",',
      '-----------^',
      '  "fullname": "Denise Pickett",'
      ].join('\n');

    // then
    const result = showFancySyntaxException(json, err);
    expect(result).toEqual(expected);
    expect(result).toMatchSnapshot();
  });

  it('case 3', () => {
    // given
    const testJson = readFixture('test3.json');
    const err = new SyntaxError('JSON.parse: bad control character in string literal at line 4 column 9 of the JSON data');

    // expected
    const expected =
      [
        'JSON.parse: bad control character in string literal at line 4 column 9 of the JSON data',
        '  "b": 5,',
        '  "c: 5',
        '-------^',
        '}'
      ].join('\n');

    // then
    const result = showFancySyntaxException(testJson, err);
    expect(result).toEqual(expected);
    expect(result).toMatchSnapshot();

  });

  it('case 4', () => {
    // given
    const testJson = readFixture('test4.json');
    const err = new SyntaxError('JSON.parse: bad control character in string literal at line 1 column 3 of the JSON data');

    // expected
    const expected =
      [
        'JSON.parse: bad control character in string literal at line 1 column 3 of the JSON data',
        '{"',
        '-^',
        '  "a": 4,'
      ].join('\n');

    // then
    const result = showFancySyntaxException(testJson, err);
    expect(result).toEqual(expected);
    expect(result).toMatchSnapshot();

  });
});
