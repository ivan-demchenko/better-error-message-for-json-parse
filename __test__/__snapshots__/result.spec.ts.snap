// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`building message case 1 1`] = `
"Unexpected string in JSON at position 118
  \\"surname\\": \\"Rowland\\"
  \\"fullname\\": \\"Denise Pickett\\",
-^
  \\"email\\": \\"jack@garrison.vc\\","
`;

exports[`building message case 2 1`] = `
"Unexpected string in JSON at position 105
  \\"name\\": \\"Eva\\",
  \\"surname\\" \\"Rowland\\",
-----------^
  \\"fullname\\": \\"Denise Pickett\\","
`;

exports[`building message case 3 1`] = `
"JSON.parse: bad control character in string literal at line 4 column 9 of the JSON data
  \\"b\\": 5,
  \\"c: 5
-------^
}"
`;

exports[`building message case 4 1`] = `
"JSON.parse: bad control character in string literal at line 1 column 3 of the JSON data
{\\"
-^
  \\"a\\": 4,"
`;
